File: app/src/main/java/com/example/coronatracker/Funtions/MyApplication.java
package com.example.coronatracker.Funtions;


import android.app.Application;
import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.content.Context;
import android.content.Intent;
import android.net.ConnectivityManager;
import android.net.NetworkInfo;
import android.os.Build;

import androidx.work.PeriodicWorkRequest;
import androidx.work.WorkManager;

import com.example.coronatracker.Activities.MainActivity;

import java.util.concurrent.TimeUnit;

public class MyApplication extends Application {
    public static final String CHANNEL_ID = "myChannel";
    public static final String KEY = "key_for_theme";
    // private FusedLocationProviderClient fusedLocationClient;
    private static MyApplication instance;
    @Override
    public void onCreate() {
        super.onCreate();
        if (instance == null) {
            instance = this;
        }
        notificationCreator();
        setupWorker();
//        fusedLocationClient = LocationServices.getFusedLocationProviderClient(this);
//        fusedLocationClient.getLastLocation()
//                .addOnSuccessListener(new OnSuccessListener<Location>() {
//                    @Override
//                    public void onSuccess(Location location) {
//
//                    }
//                })
    }

    public static Context getInstance(){
        return instance;
    }

    public static boolean hasNetwork(){
        return instance.isNetworkConnected();
    }

    private boolean isNetworkConnected() {
        ConnectivityManager cm =
                (ConnectivityManager) getSystemService(Context.CONNECTIVITY_SERVICE);

        NetworkInfo activeNetwork = cm.getActiveNetworkInfo();
        return activeNetwork != null &&
                activeNetwork.isConnectedOrConnecting();
    }

    private void notificationCreator() {

        // collapsedNotify.setTextViewText(R.id.deathsTextView,"ngo");
        if (Build.VERSION.SDK_INT > Build.VERSION_CODES.O) {
            NotificationChannel notificationChannel = new NotificationChannel(CHANNEL_ID,
                    "channel Name", NotificationManager.IMPORTANCE_DEFAULT);
            notificationChannel.setDescription("lets see how foreground works");
            NotificationManager manager = getSystemService(NotificationManager.class);
            manager.createNotificationChannel(notificationChannel);
        }
    }

    void setupWorker() {
        PeriodicWorkRequest periodicWorkRequest = new PeriodicWorkRequest
                .Builder(WorkManagerForNotifying.class, 10, TimeUnit.MINUTES)
                .build();
        WorkManager.getInstance(this).enqueue(periodicWorkRequest);
    }
}

File: app/src/main/java/com/example/coronatracker/funtions/MyApplication.kt
package com.example.coronatracker.funtions

import android.app.Application
import android.app.NotificationChannel
import android.app.NotificationManager
import android.content.Context
import android.net.ConnectivityManager
import android.os.Build
import androidx.work.*
import dagger.hilt.android.HiltAndroidApp
import java.util.concurrent.TimeUnit

@HiltAndroidApp
class MyApplication : Application() {
    override fun onCreate() {
        super.onCreate()
        if (instance == null) {
            instance = this
        }
        notificationCreator()
        setupWorker()
    }

    /**
     * @return true if connected to network
     * @return false if not connected to any network
     */
    private val isNetworkConnected: Boolean
        get() {
            val cm = getSystemService(CONNECTIVITY_SERVICE) as ConnectivityManager
            val activeNetwork = cm.activeNetworkInfo
            return activeNetwork != null &&
                    activeNetwork.isConnectedOrConnecting
        }

    private fun notificationCreator() {

        // collapsedNotify.setTextViewText(R.id.deathsTextView,"ngo");
        if (Build.VERSION.SDK_INT > Build.VERSION_CODES.O) {
            val notificationChannel = NotificationChannel(
                CHANNEL_ID,
                "channel Name", NotificationManager.IMPORTANCE_DEFAULT
            )
            notificationChannel.description = "lets see how foreground works"
            val manager = getSystemService(
                NotificationManager::class.java
            )
            manager.createNotificationChannel(notificationChannel)
        }
    }

    /**
     * setting worker for notifying User every 12 hours on behalf of his
     * @location
     */
    private fun setupWorker() {
        val constraint=Constraints.Builder()
            .setRequiresBatteryNotLow(true)
            .setRequiredNetworkType(NetworkType.CONNECTED)
            .build()
        val periodicWorkRequest = PeriodicWorkRequest.Builder(
            WorkManagerForNotifying::class.java, 12, TimeUnit.HOURS
        ).setConstraints(constraint)
            .setInitialDelay(6,TimeUnit.HOURS)
            .addTag(WORKER_TAG)
            .setBackoffCriteria(BackoffPolicy.LINEAR,
                PeriodicWorkRequest.MIN_BACKOFF_MILLIS, TimeUnit.MILLISECONDS)
            .build()
        WorkManager.getInstance().enqueueUniquePeriodicWork(
            SYNC_DATA_WORK_NAME,
            ExistingPeriodicWorkPolicy.KEEP,
            periodicWorkRequest
        )
    }

    companion object {
        const val WORKER_TAG = ""
        const val CHANNEL_ID = "myChannel"
        const val SYNC_DATA_WORK_NAME = "this is unique"

        // private FusedLocationProviderClient fusedLocationClient;
        private var instance: MyApplication? = null

        /**
         * @return instance of application
         */
        @JvmStatic
        fun getInstance(): Context? {
            return instance
        }

        /**
         * @return true if connected to network
         * @return false if not connected to any network
         */
        @JvmStatic
        fun hasNetwork(): Boolean {
            return instance!!.isNetworkConnected
        }
    }
}